crates/bevy_reflect/src/path.rs

```diff
@@ -25,395 +35,871 @@ pub enum ReflectPathError<'a> {
-                Token::Dot => {
-                    if let Some(Token::Ident(value)) = next_token(path, &mut index) {
-                        current = read_field_mut(current, value, current_index)?;
-                    } else {
-                        return Err(ReflectPathError::ExpectedIdent {
-                            index: current_index,
-                        });
-                    }
-                }
-                Token::OpenBracket => {
-                    if let Some(Token::Ident(value)) = next_token(path, &mut index) {
-                        match current.reflect_mut() {
-                            ReflectMut::List(reflect_list) => {
-                                current = read_array_entry_mut(reflect_list, value, current_index)?;
-                            }
-                            ReflectMut::Array(reflect_arr) => {
-                                current = read_array_entry_mut(reflect_arr, value, current_index)?;
-                            }
-                            _ => {
-                                return Err(ReflectPathError::ExpectedStruct {
-                                    index: current_index,
-                                })
-                            }
-                        }
-                    } else {
-                        return Err(ReflectPathError::ExpectedIdent {
-                            index: current_index,
-                        });
-                    }
+        for (access, current_index) in PathParser::new(path) {
+            current = access?.read_element_mut(current, current_index)?;
+        }
+        Ok(current)
+    }
+}
 
-                    if let Some(Token::CloseBracket) = next_token(path, &mut index) {
-                    } else {
-                        return Err(ReflectPathError::ExpectedToken {
-                            index: current_index,
-                            token: "]",
-                        });
+/// A pre-parsed path to an element within a type.
+///
+/// This struct may be used like [`GetPath`] but removes the cost of parsing the path
+/// string at each element access.
+///
+/// It's recommended to use this in place of `GetPath` when the path string is
+/// unlikely to be changed and will be accessed repeatedly.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct ParsedPath(
+    /// This is the boxed slice of pre-parsed accesses.
+    ///
+    /// Each item in the slice contains the access along with the character
+    /// index of the start of the access within the parsed path string.
+    ///
+    /// The index is mainly used for more helpful error reporting.
+    Box<[(Access, usize)]>,
+);
+
+impl ParsedPath {
+    /// Parses a [`ParsedPath`] from a string.
+    ///
+    /// See [`GetPath`] for the exact path format.
+    ///
+    /// Returns an error if the string does not represent a valid path to an element.
+    pub fn parse(string: &str) -> Result<Self, ReflectPathError<'_>> {
```

Comment:
This needs docs for what exactly a parseable path looks like, especially the # bit. Was not obvious to me at all from the tests without reading actual source.

Reply:
A nice long doc test would be great here.

Reply:
The docs regarding this is on `GetPath`, but maybe apart from just the link to those docs, I can give a couple examples here. Does that work?